UV-278 - Modbus Adapter

SammyOina opened the pull request:
### What does this do?
Creates Modbus adapter on mainflux edge

### Which issue(s) does this PR fix/relate to?
https://github.com/ultravioletrs/issues/issues/278

### List any changes that modify/break current functionality

### Have you included tests for your changes?

### Did you document any new/modified functionality?

### Notes

To test using a simulated device, 
```yaml
version: "2.1"
services:
  modbus:
    image: techplex/modbus-sim:latest
    container_name: modbus
    ports:
      - 1502:1502
    restart: unless-stopped
```
See readme for examples

SammyOina committed:
inital modbus adapter

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
add CI and issue templates

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
add http endpoint

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
add http server

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
add docker

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
add docker

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
add examples

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
add stop options

Signed-off-by: SammyOina <sammyoina@gmail.com>

rodneyosodo commented on Makefile:
```diff
@@ -0,0 +1,121 @@
+test:
+	go test -mod=vendor -v -race -count 1 -tags test $(shell go list ./... | grep -v 'vendor\|cmd')
+
+proto:
```
We don't need this here

rodneyosodo commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,20 @@
+    restart: on-failure
+    networks:
+      - mainflux-edge-base-net
+    ports:
```
Remove ports as they are already defined on the individual broker files

rodneyosodo commented on examples/publish/main.go:
```diff
@@ -0,0 +1,64 @@
+	}
+	defer ps.Close()
+
+	if err := ps.Publish(context.Background(), subj, &messaging.Message{
```
define message before publishing
```go
var msg = &messaging.Message{}
if err := ps.Publish(context.Background(), subj, msg)
```

rodneyosodo commented on examples/publish/main.go:
```diff
@@ -0,0 +1,64 @@
+}
+
+func usage() {
+	log.Printf("Usage: publish [-s server] <channel> <msg>\n")
```
Add example of contents for `<channel>` and `<msg>`

rodneyosodo commented on examples/subscribe/main.go:
```diff
@@ -0,0 +1,105 @@
+		return
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
```
Define them at the start

rodneyosodo commented on go.mod:
```diff
@@ -0,0 +1,83 @@
+
+go 1.21.0
+
+replace github.com/mainflux/mainflux => github.com/mainflux/mainflux v0.0.0-20230811131717-9dbe87f03a1a
```
replace

SammyOina replied:
what do you mean

rodneyosodo replied:
rather remove it. Why do we need a specific version of mainflux?

arvindh123 commented on modbus/api/endpoint.go:
```diff
@@ -0,0 +1,91 @@
+				return nil, err
+			}
+
+			if err = pub.Publish(ctx, fmt.Sprintf("modbus.read.tcp.%s", req.Datapoint), &messaging.Message{Payload: payload}); err != nil {
+				return nil, err
```
Topic is hard coded. But this also raises me a question about how the data flows from edge to Mainflux , 
So We need to dicsuss about the data flow.

SammyOina committed:
Refactor modbus adapter to use rpc

Refactor the modbus adapter codebase by removing unnecessary imports and unused constants.

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
Fix Makefile test command and remove unnecessary proto generation

The Makefile test command was modified to remove the "-mod=vendor" flag and the proto generation commands were removed as they are unnecessary.

This commit fixes the Makefile test command and removes unnecessary proto generation commands.

Signed-off-by: SammyOina <sammyoina@gmail.com>

rodneyosodo commented on cmd/modbus/main.go:
```diff
@@ -0,0 +1,96 @@
+		return server.Start(ctx)
+	})
+
+	fmt.Printf("modbus service listening on rpc %s:%d", cfg.RPCHost, cfg.RPCPort)
```
Use logger

rodneyosodo commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+version: "3.7"
+
```
Do we need this?

SammyOina replied:
Agent and Export will be moving to this repo so yes

rodneyosodo commented on modbus/modbus.go:
```diff
@@ -0,0 +1,266 @@
+var _ ModbusService = (*modbusService)(nil)
+
+// adapterService provides methods for reading and writing data on Modbus.
+type modbusService struct {
```
```suggestion
type service struct {
```

Also fix the comment

rodneyosodo commented on modbus/setup_test.go:
```diff
@@ -0,0 +1,60 @@
+)
+
+var (
+	Address string
```
```suggestion
	address string
```

rodneyosodo commented on modbus/modbus.go:
```diff
@@ -0,0 +1,266 @@
+		handler.Timeout = config.Timeout.Duration
+	}
+
+	if err := handler.Connect(); err != nil {
```
repeated

rodneyosodo commented on modbus/adapter.go:
```diff
@@ -0,0 +1,97 @@
+	Write(config RWOptions, res *[]byte) error
+	// Configure sets the configuration for a device and returns the index for the connection.
+	ConfigureTCP(config TCPHandlerOptions, id *int) error
+	ConfigureRTU(config RTUHandlerOptions, id *int) error
```
godocs for this

rodneyosodo commented on modbus/README.md:
```diff
@@ -0,0 +1,163 @@
+| MF_MODBUS_ADAPTER_LOG_LEVEL        | Service log level                        | info                           |
+| MF_JAEGER_URL                      | Jaeger server URL                        | http://jaeger:14268/api/traces |
+| MF_MODBUS_ADAPTER_INSTANCE_ID      | Modbus adapter instance ID               |                                |
+| MF_MODBUS_ADAPTER_RPC_HOST        | Modbus service HTTP host                 |                                |
```
align

rodneyosodo commented on modbus/README.md:
````diff
@@ -0,0 +1,163 @@
+
+```bash
+# download the latest version of the service
+git clone https://github.com/mainflux/mainflux
````
change

rodneyosodo commented on modbus/README.md:
````diff
@@ -0,0 +1,163 @@
+To start reading values, you need to perform an RPC call to the read method.
+
+```go
+configRead := modbus.RWOptions{
````
align indentation

rodneyosodo commented on modbus/README.md:
````diff
@@ -0,0 +1,163 @@
+To start writing values, you need to perform an RPC call to the write method.
+
+```go
+configWrite := modbus.RWOptions{
````
align indentation

rodneyosodo commented on Makefile:
```diff
@@ -0,0 +1,116 @@
+	done
+	$(call docker_push,$(version))
+
+rundev:
```
add dev scripts for this to work

SammyOina committed:
Fix logger output in modbus service

The logger output in the modbus service was not being properly formatted. This commit fixes the issue by using the logger.Info function to print the modbus service listening information.

- Update logger output in modbus service to use logger.Info function
- Fix formatting issue with logger output in modbus service

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
Fix subscription and publishing bugs

This commit fixes bugs related to subscription and publishing in the examples/publish/main.go and examples/subscribe/main.go files. The changes include adding the context package and passing the context to the Publish and Subscribe functions. This ensures that the functions have the necessary context for proper execution.

The bug in the examples/publish/main.go file is fixed by adding the context package and passing the context to the Publish function. Previously, the Publish function was called without a context, which could lead to unexpected behavior.

The bug in the examples/subscribe/main.go file is fixed in a similar manner. The context package is imported, and the context is passed to the Subscribe function. This ensures that the Subscribe function has the necessary context for proper execution.

These changes ensure that the subscription and publishing functionalities work correctly in the examples.

Signed-off-by: SammyOina <sammyoina@gmail.com>

rodneyosodo commented on modbus/adapter.go:
```diff
@@ -0,0 +1,98 @@
+type Service interface {
+	// Read subscribes to the Subscriber and
+	// reads modbus sensor values while publishing them to publisher.
+	Read(config RWOptions, res *[]byte) error
```
Why not
```suggestion
	Read(config RWOptions) (*[]byte, error)
```

SammyOina replied:
its an RPC function

rodneyosodo commented on internal/apiutil/transport.go:
```diff
@@ -0,0 +1,152 @@
+	"github.com/mainflux/mainflux/logger"
+	"github.com/mainflux/mainflux/pkg/errors"
+)
+
```
Remove unused

rodneyosodo commented on internal/apiutil/token.go:
```diff
@@ -0,0 +1,37 @@
+
+// ThingPrefix represents the key prefix for Thing authentication scheme.
+const ThingPrefix = "Thing "
+
```
Remove unused

rodneyosodo commented on internal/apiutil/errors.go:
```diff
@@ -0,0 +1,146 @@
+// Errors defined in this file are used by the LoggingErrorEncoder decorator
+// to distinguish and log API request validation errors and avoid that service
+// errors are logged twice.
+var (
```
Remove unused

SammyOina committed:
remove unused code

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
remove vendor

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
remove jaeger

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
remove unused code

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
remove internal

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
remove unused code

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
update docs

Signed-off-by: SammyOina <sammyoina@gmail.com>
